### PR TITLE
feat(core): support TypeScript 5.1

### PR DESCRIPTION
**Pull request by crisbeto**
Updates the project to support building with TypeScript 5.1.

### CODE REVIEW

File: goldens/public-api/common/http/index.md [hunk 1]
```diff
@@ -1917,7 +1917,7 @@ export class HttpRequest<T> {
     readonly params: HttpParams;
     readonly reportProgress: boolean;
     readonly responseType: 'arraybuffer' | 'blob' | 'json' | 'text';
-    serializeBody(): ArrayBuffer | Blob | FormData | string | null;
+    serializeBody(): ArrayBuffer | Blob | FormData | URLSearchParams | string | null;
```
Comment (crisbeto) [inline goldens/public-api/common/http/index.md:1920]: While this is technically a public API change, we've been returning `URLSearchParams` [for about two years](https://github.com/angular/angular/pull/37852). The type just never reflected it.

**Review by josephperrott — APPROVED** [review]
LGTM

**Comment by josephperrott** [comment]
Added a blocked state because we need to be sure that we don't accidentally before TS 5.1 is final

**Comment by crisbeto** [comment]
@josephperrott we need to merge this in so the other repositories can update. This is targeted for the next minor release which comes out after our next minor.

File: package.json [hunk 2]
```diff
@@ -154,7 +154,7 @@
     "tsickle": "0.46.3",
     "tslib": "^2.3.0",
     "tslint": "6.1.3",
-    "typescript": "5.0.2",
+    "typescript": "5.1.0-beta",
```
Comment (JoostK) [inline package.json:157]: Living on the edge! 🎢

**Review by JoostK — APPROVED** [review]
That must've been the smoothest TS upgrade I have ever seen.

**Review by alxhub — APPROVED** [review]

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 69dadd25020ee84364466c0740f695984dd8c84d.

**Comment by e-oz** [comment]
From 16.0.1:

Error: Failed to initialize Angular compilation - The Angular Compiler requires TypeScript >=4.9.3 and <5.1.0 but 5.1.0-dev.20230511 was found instead

🙀

**Comment by JoostK** [comment]
> From 16.0.1:
> 
> Error: Failed to initialize Angular compilation - The Angular Compiler requires TypeScript >=4.9.3 and <5.1.0 but 5.1.0-dev.20230511 was found instead
> 
> 🙀

This is an Angular 16.1 feature.

**Comment by e-oz** [comment]
You are right, sorry, missread the changelog

**Pull request closed without merging**